subscribe/index.ts

```diff
@@ -0,0 +1,70 @@
+import { useEffect } from 'react'
+import useSWR from 'swr'
+import { withMiddleware } from '../src/utils/with-middleware'
+import { Key, SWRHook, Middleware, SWRConfiguration } from '../src/types'
+
+export type SWRSubscription<Data = any, Error = any> = (
+  key: Key,
+  callbacks: {
+    onData(data: Data): void
+    onError(err: Error): void
+  }
+) => void
+
+export type SWRSubscriptionResponse<Data = any, Error = any> = {
+  data?: Data
+  error?: Error
+}
+
+export type SWRSubscriptionHook<Data = any, Error = any> = (
+  key: Key,
+  subscribeFn: SWRSubscription<Data, Error>,
+  config?: SWRConfiguration
+) => SWRSubscriptionResponse<Data, Error>
+
+const subscriptions = new Map<Key, number>()
+const disposers = new Map()
+
+export const subscribe = (<Data, Error>(useSWRNext: SWRHook) =>
+  (
+    key: Key,
+    subscribeFn: SWRSubscription<Data, Error>,
+    config?: SWRConfiguration
+  ): SWRSubscriptionResponse<Data, Error> => {
+    const { data, error, mutate } = useSWRNext(key, null, config)
+
+    useEffect(() => {
+      subscriptions.set(key, (subscriptions.get(key) || 0) + 1)
+      const onData = (val: Data) => mutate(val, false)
+      const onError = async (err: any) => {
+        // Avoid thrown errors from `mutate`
+        // eslint-disable-next-line no-empty
+        try {
+          await mutate(() => {
+            throw err
+          }, false)
+        } catch (_) {
+          /* eslint-disable-line no-empty */
+        }
+      }
+
+      if (subscriptions.get(key) === 1) {
+        const dispose = subscribeFn(key, { onData, onError })
+        disposers.set(key, dispose)
+      }
+      return () => {
```

Comment:
I’d wrap the cleanup logic with setTimeout(..., 0) to prevent unmount-mount scenarios caused by state updates. Think of 2 tabs with the same subscription hooks inside, but we only render the active one. If you switch the tab the subscription will be unsubscribed and then resubscribed. This is a common thing we can improve by delaying it til the next tick. 

Also, for strict effects it will be double invoked so this is a nice trick to avoid unnecessary network requests.